feat: anon piece selection and retrieval (#487)
FilOzone/dealbot: docs/checks/anon-retrievals.md

### Definition of Successful Retrieval

A successful anonymous retrieval requires:

1. **Piece fetch** — `GET {spBaseUrl}/piece/{pieceCid}` returns HTTP 2xx and the response bytes hash to the declared CommP (piece CID).

If the piece advertises IPFS indexing (`withIPFSIndexing = true` and a non-null `ipfsRootCid`), three additional dimensions are validated *independently*. Importantly, they do not gate the overall `piece_fetch_status`, and each is recorded as its own outcome column / metric:

2. **CAR parseable:** the fetched bytes parse as a CAR file.
3. **IPNI:** the SP is advertised as a provider for the root CID and a sample of child CIDs via filecoinpin.contact.
4. **Block fetch:** a sample of CIDs from the parsed CAR is re-fetched via `{spBaseUrl}/ipfs/{cid}?format=raw` and each response is hash-verified against its declared CID.

A piece without IPFS indexing is exercised only at step (1).

Operational timeouts exist to prevent jobs from running indefinitely. If the job exceeds `ANON_RETRIEVAL_JOB_TIMEOUT_SECONDS`, it is aborted; a row is still emitted so that partial metrics (TTFB, bytes, response code) are not lost.

## Piece Selection

> Unlike the [Retrieval check](./retrievals.md#piece-selection), dealbot does not retrieve from its own deals. Pieces are sampled from the on-chain subgraph of all FWSS-served pieces for the SP under test.

Reviewer: ```suggestion
Unlike the [Retrieval check](./retrievals.md#piece-selection), dealbot does not retrieve from its own deals. Pieces are sampled from the [on-chain subgraph](../../src/subgraph) of all FWSS-served pieces for the SP under test.
```

   - `small` (1–20 MiB) — 20%
   - `medium` (20–100 MiB) — 50%
   - `large` (100–500 MiB) — 30%
2. **Pick a pool**:
   - `indexed` (IPFS-indexed pieces) — 80%
   - `any` (all FWSS pieces) — 20%
3. **Generate a uniform-random `sampleKey`** and query the subgraph for the smallest `Root.sampleKey ≥ $sampleKey` matching the SP, payer, size range, and pool filters.
4. **Drop the candidate** if `pdpPaymentEndEpoch` has passed.
5. **Fall back** through: (same bucket, opposite pool) → (any bucket, indexed) → (any bucket, any).

The 80/20 split for `indexed` vs `any` exists so that SPs cannot optimize only their CAR corpus and still appear healthy on this check.

> [!NOTE]
> The bucket sizes were chosen such that the whole file will still fit into memory. In the future we may implement a streaming verification and parsing.

Source: [`anon-piece-selector.service.ts`](../../apps/backend/src/retrieval-anon/anon-piece-selector.service.ts)

## What Happens Each Cycle

```mermaid
flowchart TD
  Select["Sample anonymous piece for SP from subgraph"] --> Fetch["GET /piece/{pieceCid}"]
  Fetch --> CommP["Hash bytes → verify CommP"]
  CommP --> HasIpfs{"piece.withIPFSIndexing<br/>and ipfsRootCid?"}

>   HasIpfs -- "no" --> Record["Persist row + metrics"]

Reviewer: ```suggestion
  HasIpfs -- "no" --> Record["Persist Clickhosue row + emit Prometheus metrics"]
```

I know that's long and I'm not wedded to it, but I am wondering for making it clear what's happening.

  HashCheck --> Record
```

### Piece Fetch

- **URL:** `{spBaseUrl}/piece/{pieceCid}` (HTTP/2)
- **Buffered in memory** — piece sizes are capped at 500 MiB by selection.
- **Validates CommP** — the CommP of the response bytes must match `pieceCid`.

Source: [`piece-retrieval.service.ts`](../../apps/backend/src/retrieval-anon/piece-retrieval.service.ts)

### CAR Validation (only when piece advertises IPFS indexing)

When the selected piece has `withIPFSIndexing = true` and a non-null `ipfsRootCid`, the fetched bytes are parsed as a CAR and a random sample of `ANON_RETRIEVAL_BLOCK_SAMPLE_COUNT` CIDs is exercised:

- **IPNI check:** `IpniVerificationService.verify(rootCid, sampledCids, sp)` polls filecoinpin.contact until each CID resolves to the SP under test, the timeout fires, or `IPNI_VERIFICATION_TIMEOUT_MS` is reached.
- **Block fetch check:** for each sampled CID, fetch `{spBaseUrl}/ipfs/{cid}?format=raw` and hash-verify the response against the CID. Non-2xx, hash mismatch, unsupported codec, or transport errors all count as a single failed block.

Source: [`car-validation.service.ts`](../../apps/backend/src/retrieval-anon/car-validation.service.ts)

## What Gets Asserted

> | # | Assertion | How It's Checked | Retries | Relevant Metric | Implemented? |

Reviewer: I think we can drop the implemented column given everything is implemented.

I think we only need that if we haven't implementing certain things. (That is a legacy remnant from our our initial docs where we were specing out what to do but hadn't implemented it all yet.)

Author: Done in https://github.com/FilOzone/dealbot/pull/487/changes/2850db821e2a340a28c4c3b0a9645fbc4ffb21d2

| 3 | Bytes parse as a CAR (IPFS-indexed pieces only) | `@ipld/car` parses the response | 0 | [`anonCarParseStatus`](./events-and-metrics.md#anonCarParseStatus) | Yes |
| 4 | SP is advertised on IPNI for root + sampled CIDs | filecoinpin.contact returns provider records | polling until timeout | [`anonIpniStatus`](./events-and-metrics.md#anonIpniStatus) | Yes |
| 5 | Sampled blocks fetch + hash-verify | `/ipfs/{cid}?format=raw` for each sample | 0 | [`anonBlockFetchStatus`](./events-and-metrics.md#anonBlockFetchStatus) | Yes |

## Result Recording

Each anonymous retrieval attempt writes one row to the `anon_retrieval_checks` ClickHouse table. The row is emitted **even on abort or unexpected error** so that the partial evidence (TTFB, bytes, response code) is preserved.

The DDL and column-level comments in [`clickhouse.schema.ts`](../../apps/backend/src/clickhouse/clickhouse.schema.ts) are authoritative. The summary below is for orientation.

| Column | Meaning |
|--------|---------|
| `timestamp` | When the check started (ms UTC) |

> | `probe_location` | Dealbot probe location (`DEALBOT_PROBE_LOCATION`) |

Reviewer: Is this defined more?

Author: This isn't defined in this PR. There are some unresolved discussions in https://github.com/FilOzone/dealbot/issues/479 and https://github.com/FilOzone/dealbot/issues/246.

If okay, I'd defer the documentation until there's consensus and these issues are resolved?

Reviewer: Ok, got it. So I think we either:
1. Don't have this column for now until we have #246 OR
2. Indicate this isn't used currently but will be updated to be used as part of #246

Reviewer: I assume we'll go with approach number 2 (add a comment linking to backlog item)

Author: Oh sorry, I skipped over this comment. I added two notes in https://github.com/FilOzone/dealbot/pull/487/changes/4666e0cc8056bedcdd663bdca06f7d0091a93ea7

2. **CAR parseable:** the fetched bytes parse as a CAR file.
3. **IPNI:** the SP is advertised as a provider for the root CID and a sample of child CIDs via filecoinpin.contact.
4. **Block fetch:** a sample of CIDs from the parsed CAR is re-fetched via `{spBaseUrl}/ipfs/{cid}?format=raw` and each response is hash-verified against its declared CID.

A piece without IPFS indexing is exercised only at step (1).

Operational timeouts exist to prevent jobs from running indefinitely. If the job exceeds `ANON_RETRIEVAL_JOB_TIMEOUT_SECONDS`, it is aborted; a row is still emitted so that partial metrics (TTFB, bytes, response code) are not lost.

## Piece Selection

Unlike the [Retrieval check](./retrievals.md#piece-selection), dealbot does not retrieve from its own deals. Pieces are sampled from the [on-chain subgraph](../../src/subgraph) of all FWSS-served pieces for the SP under test.

Selection strategy (per scheduled job, per SP):

1. **Pick a size bucket** by weighted random:
   - `small` (1–20 MiB) — 20%
   - `medium` (20–100 MiB) — 50%

>    - `large` (100–500 MiB) — 30%

Reviewer: This needs updating right?

Source: [`car-validation.service.ts`](../../apps/backend/src/retrieval-anon/car-validation.service.ts)

## What Gets Asserted

| # | Assertion | How It's Checked | Retries | Relevant Metric |
|---|-----------|------------------|:---:|------------------|
| 1 | SP serves the piece | `GET /piece/{pieceCid}` returns HTTP 2xx | 0 | [`anonPieceRetrievalLastByteMs`](./events-and-metrics.md#anonPieceRetrievalLastByteMs) |
| 2 | Bytes match the declared CommP | Hash of response bytes equals `pieceCid` | 0 | [`anonPieceRetrievalStatus`](./events-and-metrics.md#anonPieceRetrievalStatus) |
| 3 | Bytes parse as a CAR (IPFS-indexed pieces only) | `@ipld/car` parses the response | 0 | [`anonCarParseStatus`](./events-and-metrics.md#anonCarParseStatus) |
| 4 | SP is advertised on IPNI for root + sampled CIDs | filecoinpin.contact returns provider records | polling until timeout | [`anonIpniStatus`](./events-and-metrics.md#anonIpniStatus) |
| 5 | Sampled blocks fetch + hash-verify | `/ipfs/{cid}?format=raw` for each sample | 0 | [`anonBlockFetchStatus`](./events-and-metrics.md#anonBlockFetchStatus) |

## Sub-status meanings

> Unlike a Data Storage deal, anonymous retrieval does **not** have a single rolled-up status. CAR / IPNI / block-fetch outcomes are recorded independently and do **not** flip the piece-fetch verdict. Each status metric below is emitted exactly once per check (except `anonPieceRetrievalStatus`, which is also emitted as `failure.no_piece` when piece selection itself fails).

Reviewer: ```suggestion
Unlike the [Data Storage check](./data-storage.md#deal-status-progression), anonymous retrieval does **not** have a rolled-up status (e.g., `anonRetrievalStatus). Piece retrieval, CAR parsing, IPNI verification, block-fetch outcomes are recorded independently. Each status metric below is emitted exactly once per check, except when `anonPieceRetrievalStatus=failure.no_piece` because selection itself fails.
```

At least for me that is more clear.

That said, seeing this, I'm wondering why again why we're deviating from the data storage check and doming more "skipped" status values. I think we should either make these consistent, or add some explanation for why we're deviating. And if we covered this and I'm just forgetting this, my apologies!

Author: I'd say because we have some conditional checks in the anon retrieval code paths which somehow need to be recorded in the database. For these I'd clearly mark parts of the retrieval check as skipped if they didn't run (an alternative would indeed be a null value).

| anonPieceRetrievalStatus | Meaning |
|--------|---------|
| `success` | `GET /piece/{pieceCid}` returned HTTP 2xx **and** the response bytes hashed to the declared CommP. |
| `failure.http` | Piece fetch did not return HTTP 2xx, or the request failed at the transport layer (DNS, TLS, connection reset, etc.). |
| `failure.commp` | Piece fetch returned HTTP 2xx, but the response bytes hashed to a different CID than `pieceCid`. The bytes are discarded — downstream CAR / IPNI / block-fetch validation is skipped to avoid amplifying a misbehaving SP. |
| `failure.timedout` | The job-level `AbortSignal` fired (most often `ANON_RETRIEVAL_JOB_TIMEOUT_SECONDS`). Partial timing/byte evidence is still persisted. |

> | `failure.no_piece` | The subgraph returned no candidate piece for the SP after all selection fallbacks. No HTTP request was attempted. |

Reviewer: Given this isn't the SP's "fault" per se, maybe this should be listed a failure? Do we instead call this status as `skipped` to be consistent with other `skipped` usage below.

Also, maybe put this status below `success` and before the other `failure`s?

Author: > maybe this should be listed a failure?

I guess you mean "should **_not_** be listed as failure"? Anyhow, I changed it to `skipped` and moved it 👍

| `failure.other` | Catch-all for retrieval failures that do not match any of the categories above. |

| anonCarParseStatus | Meaning |
|--------|---------|
| `parseable` | The fetched piece bytes were successfully parsed as a CAR by `@ipld/car`. |
| `not_parseable` | The fetched piece bytes could not be parsed as a CAR (malformed header, truncated content, unexpected encoding, or parser threw). |

> > Emitted only when piece fetch succeeded **and** the piece advertises IPFS indexing (`withIPFSIndexing = true` with a non-null `ipfsRootCid`). Skipped otherwise; no row value is recorded.

Reviewer: A few things:
1. In rendered view, it's not clear which status this applies to (carParse or ipni). 
2. (nit) Maybe use github `> [!NOTE]` syntax?
3. Higher level: why don't we just have a `anonCarParseStatus=skipped`? Why are we treating `anonCarParseStatus` special?

Author: See comment: https://github.com/FilOzone/dealbot/pull/487#issuecomment-4553784144

| anonIpniStatus | Meaning |
|--------|---------|
| `valid` | filecoinpin.contact returned the SP as a provider for the root CID **and** every sampled child CID within `IPNI_VERIFICATION_TIMEOUT_MS`. |
| `invalid` | IPNI was queried but at least one CID never resolved to the SP under test before the timeout (or the timeout fired with unresolved CIDs). |

> | `skipped` | IPNI verification was not attempted — piece fetch failed, the piece does not advertise IPFS indexing, or the SP is not registered with `WalletSdkService` so no IPNI sampling could run. |

Reviewer: > the SP is not registered with `WalletSdkService` so no IPNI sampling could run.

How could this even occur?
(Also as someone just reading the "checks" docs, I don't think I know what the `WalletSdkService` means.)

Author: yeah, right. It's actually an invariant in the system that the SP is always registered. It's just a defensive code path that would emit an error if for whatever reason the SP isn't registered. I'll remove the `WalletSdkService ` reference from here to avoid confusion.

> | `error` | IPNI verification was attempted but the CAR-validation step threw before producing a result (e.g. invalid root CID, transport error, unexpected exception). |

Reviewer: This seems a bit weird to me, that an invalid root CID or car-parsing exception would cause this status to switch to error. I would expect to just be `skipped`.

I would expect we don't even start IPNI verification until we have validated the CAR. That is is what's in the flow diagram. I guess we can do IPNI verification for the root CID before doing the car validation, but that also seems like unnecessary complexity. We can get the CAR, make sure the car is valid, and then create a set of CIDs to verify with IPNI (including the root CID).

Author: See comment: https://github.com/FilOzone/dealbot/pull/487#issuecomment-4553784144

| anonBlockFetchStatus | Meaning |
|--------|---------|
| `valid` | Every sampled CID was fetched via `GET {spBaseUrl}/ipfs/{cid}?format=raw` and the response bytes hash-verified against the declared CID. |
| `invalid` | At least one sampled block fetch failed: non-2xx HTTP, hash mismatch, unsupported codec, unsupported hash, or transport error. Each failed sample counts as one failed block. |
| `skipped` | Block-fetch sampling was not attempted — piece fetch failed, the piece does not advertise IPFS indexing, or CAR parsing produced no sampleable CIDs. |

> | `error` | Block-fetch sampling was attempted but the CAR-validation step threw before completing (e.g. CAR parser threw, unexpected exception). |

Reviewer: Similar to above, I wouldn't even attempt to fetch blocks until we successfully parse the car. If that fails, then we should just skip.

Author: see comment: https://github.com/FilOzone/dealbot/pull/487#issuecomment-4553784144

Source: [`piece-validation.service.ts`](../../apps/backend/src/retrieval-anon/piece-validation.service.ts)

## What Gets Asserted

| # | Assertion | How It's Checked | Retries | Relevant Metric |
|---|-----------|------------------|:---:|------------------|
| 1 | SP serves the piece | `GET /piece/{pieceCid}` returns HTTP 2xx | 0 | [`anonPieceRetrievalLastByteMs`](./events-and-metrics.md#anonPieceRetrievalLastByteMs) |
| 2 | Bytes match the declared CommP | Hash of response bytes equals `pieceCid` | 0 | [`anonPieceRetrievalStatus`](./events-and-metrics.md#anonPieceRetrievalStatus) |
| 3 | Bytes parse as a CAR (IPFS-indexed pieces only) | `@ipld/car` parses the response | 0 | [`anonCarParseStatus`](./events-and-metrics.md#anonCarParseStatus) |
| 4 | SP is advertised on IPNI for root + sampled CIDs | filecoinpin.contact returns provider records | polling until timeout | [`anonIpniStatus`](./events-and-metrics.md#anonIpniStatus) |
| 5 | Sampled blocks fetch + hash-verify | `/ipfs/{cid}?format=raw` for each sample | 0 | [`anonBlockFetchStatus`](./events-and-metrics.md#anonBlockFetchStatus) |

## Sub-status meanings

> Unlike the [Data Storage check](./data-storage.md#deal-status-progression), anonymous retrieval does **not** have a rolled-up status (e.g., `anonRetrievalStatus). Piece retrieval, CAR parsing, IPNI verification, block-fetch outcomes are recorded independently. Each status metric below is emitted exactly once per check, except when `anonPieceRetrievalStatus=skipped` because selection itself failed and no HTTP/CAR/IPNI/block-fetch work ran.

Reviewer: ```suggestion
Unlike the [Data Storage check](./data-storage.md#deal-status-progression), anonymous retrieval does **not** have a rolled-up status (e.g., `anonRetrievalStatus`). Piece retrieval, CAR parsing, IPNI verification, block-fetch outcomes are recorded independently. Each status metric below is emitted exactly once per check, except when `anonPieceRetrievalStatus=skipped` because selection itself failed and no HTTP/CAR/IPNI/block-fetch work ran.
```

| `parseable` | The fetched piece bytes were successfully parsed as a CAR by `@ipld/car`.                                                                             |
| `not_parseable` | The fetched piece bytes could not be parsed as a CAR (malformed header, truncated content, unexpected encoding, or parser threw an error). |
| `skipped` | CAR parsing was not attempted — piece fetch failed, the piece does not advertise IPFS indexing, or the job aborted before parsing.                    |

| anonIpniStatus | Meaning |
|--------|---------|
| `valid` | filecoinpin.contact returned the SP as a provider for the root CID **and** every sampled child CID within `IPNI_VERIFICATION_TIMEOUT_MS`. |
| `invalid` | IPNI was queried but at least one CID never resolved to the SP under test before the timeout (or the timeout fired with unresolved CIDs). |
| `skipped` | IPNI verification was not attempted — piece fetch failed, the piece does not advertise IPFS indexing, CAR parsing returned `not_parseable`, the root CID itself failed to parse, or the job aborted. |

> | `error` | IPNI verification was attempted and `IpniVerificationService.verify` threw unexpectedly (transport error, service down, etc.). |

Reviewer: My head is admittedly hurting from looking our metrics documentation today for other PRs. This was probably covered, but why are we coming up with different status names?

In some other ipni status work today (https://github.com/FilOzone/dealbot/pull/581 ), I settled on just using success, failure.timedout, failure.other. See cid.contact verification at https://github.com/FilOzone/dealbot/blob/8ae54af8bf63b46ed211e95bb8c9e96f6b945dcc/docs/checks/data-storage.md#sub-status-meanings

Author: Hi @BigLep no worries! I don't think it was covered as I didn't look closely enough how things were done left and right.

I aligned all sub status codes with the `success` `failure.*` convention in https://github.com/FilOzone/dealbot/pull/487/changes/b3d7731f771a0f2908d5c220874001b4d02beb94.

If you could check this commit in isolation again?

To summarize:

- aligned sub status codes to the data-storage vocabulary (success / failure.* / skipped).
- `anonCarParseStatus` → success/skipped/failure.not_parseable
- `anonIpniStatus` → success/skipped/failure.timedout/failure.other
- `anonBlockFetchStatus` → success/skipped/failure.other - dropped the `ERROR` case and collapsed it into `failure.other` because I think it's semantically the same. For context: I previously had two distinct error cases because they were raised from two different code paths.
- Piece fetch keeps failure.http/failure.commp as failure.* sub-cases

Reviewer: https://github.com/FilOzone/dealbot/commit/b3d7731f771a0f2908d5c220874001b4d02beb94 looks good. Resolving this conversation.
